### ISSUE TITLE
x/tools/gopls: revisit the set of 'quickfix' refactorings

### ISSUE THREAD

**Issue by alanlwb** [golang/go#71038]
Why can gopls@v0.16.2 do fill struct code action and gopls@v0.17.1 doesn't

feature or bug ?
<img width="633" alt="image" src="https://github.com/user-attachments/assets/82e11a15-603e-4eee-a4b1-0f2dababe2b1" />
gopls@v0.17.1 

<img width="311" alt="image" src="https://github.com/user-attachments/assets/7976b879-e6d9-4ae2-a7c4-fca583b9ba4d" />
gopls@v0.16.2

**Comment by findleyr** [comment]
Thanks @h9jiang for the analysis.

Sounds like we need to revisit the set of refactorings available in the quickfix menu, or consider changes to the VS Code extension to make refactorings more readily available.

**Comment by findleyr** [comment]
@igor-spirin-indriver yes, we'll fix this for gopsl@v0.18.0 (which is actually quite soon as we want to release this before Go 1.24).

We think we interpreted 'quick fix' too literally, only returning fixes for diagnostics.
How about the following heuristics:

- Quick fixes must be ~guaranteed to succeed (one of the main complaints that led us to constrain fixes was that 'inline this call' was provided in the quick fix menu but would frequently fail).
- Quick fixes must not break the build. This should really be a subset of the previous bullet, but must unfortunately be called out independently as not all of our code actions leave the workspace in a working state.
- Quick fixes must be frequently useful. This is of course a gray area, but we can use telemetry to determine which code actions our users are regularly invoking. I think 'generative' quick fixes, such as fill struct, are likely to be used frequently, as they are invoked as part of writing new code. Therefore, we want them to be easily accessible.

@adonovan @h9jiang what do you think? If we agree, we can document these heuristics and fix our dispatch.

**Comment by imkos** [comment]
we want to be able to display the lightbulb (automatic) when the struct can be filled. Let's fill this struct more quickly.
